src/proto/streams/recv.rs

```diff
@@ -562,17 +567,27 @@ impl<B, P> Recv<B, P>
                 unimplemented!();
             }
             None => {
-                if stream.state.is_recv_closed() {
-                    // There will be no trailer frame
-                    Ok(None.into())
-                } else {
-                    // Request to get notified once another frame arrives
-                    stream.recv_task = Some(task::current());
-                    Ok(Async::NotReady)
-                }
+                self.schedule_recv(stream)
             }
         }
     }
+
+    fn schedule_recv<T>(&mut self, stream: &mut Stream<B, P>)
+        -> Poll<Option<T>, proto::Error>
+    {
+        match stream.state.ensure_recv_open() {
+            Ok(()) => {
+                // Request to get notified once more frames arrive
+                stream.recv_task = Some(task::current());
+                Ok(Async::NotReady)
+            },
+            Err(proto::Error::Proto(Reason::NoError)) => {
```

Comment:
Could you explain this logic? I'm not sure I follow when `ensure_recv_open` would return `NoError` and why this is mapped as an end of stream.

Reply:
This would return `NoError` when either half closed, or closed with no error. This is treating the `NoError` variant similar to how IO treats the `WouldBlock` variant: for the purposes of this function, it's not an error, but means the stream is closed.

Alternatively, I can change `ensure_recv_open` to return a `Poll<(), Error>`, and convert `NoError` into an `Ok(NotReady)` in that function, and just pass it along here.